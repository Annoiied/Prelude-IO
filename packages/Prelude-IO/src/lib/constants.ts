export const VALIDATORS_KEY = Symbol("@Prelude-IO/VALIDATORS_FIELD");
