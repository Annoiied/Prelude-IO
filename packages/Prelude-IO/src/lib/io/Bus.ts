import { Tuple2, Vector } from "prelude-ts";
import { IODecode, IOAsyncDecode, IOLeft } from "./types";

const operatorRegex = /(\|\||->)/g;

/**
 * Wraps a name in parentheses if it contains an operator.
 * Might need improvements in the future regarding presedence and pre-wrapped names.
 *
 * @param {string} name The name to check for operators
 * @returns {string} The name wrapped in parentheses if it contains an operator
 */
const maybeWrap = (name: string): string =>
  operatorRegex.test(name) ? `(${name})` : name;

/**
 * Merges the names using -> or || depending on the mode.
 *
 * Wraps the names in parentheses if they contain an operator.
 *
 * @example mergeNames(["a", "b"], "->") === "a -> b"
 * @example mergeNames(["a", "b -> c"], "||") === "a || (b -> c)"
 *
 * @param {Tuple2<string, string>} names A tuple of names to merge
 * @param {"->" | "||"} mode The mode to use for merging (and/or)
 *
 * @returns {string} The merged names
 */
export const mergeNames = (names: Tuple2<string, string>, mode: "->" | "||") =>
  names.toArray().map(maybeWrap).join(` ${mode} `);

/**
 * Merges the errors of two results. Works for both left and right results.
 *
 * @param {IOLeft} us The first result
 * @param {IOLeft} them The second result
 *
 * @returns {IOLeft} A result containing the errors of both results (if any)
 */
export const mergeLeft = (us: IOLeft, them: IOLeft) =>
  us.mapLeft((usErrors) =>
    them.getLeftOrElse(Vector.empty()).appendAll(usErrors)
  ) as IOLeft;

/**
 * A bus is a wrapper around a validator that allows for chaining of validators.
 *
 * Busses are executed in parallel, so the order of the errors is not guaranteed (but should not matter).
 * When chaining busses, the decoded value of the first bus is returned, the others are ignored. This might change in the future.
 *
 * Validators do not need to be asyncronous on initiation for ease of use, but on Bus creation, they will
 * always be wrapped in a Promise.
 *
 * Promise validators should ALWAYS resolve to a Result, never throw an error or reject. This is a philosophical choice based on
 * functional programming paradigms. Result objects allow functions to explicitly return errors, and the caller can decide
 * what to do with them. Throwing errors is a side effect, and should be avoided.
 *
 * @typeparam I The input type of the bus
 * @typeparam O The output type of the bus
 *
 * @field name The name of the bus
 * @field decode The decode function.
 */
export default class Bus<I, O> {
  private constructor(
    public readonly name: string,
    public readonly decode: IOAsyncDecode<I, O>
  ) {
    Object.freeze(this);
  }

  /**
   * Creates a new bus with the given name and validator. Wraps the validator in an async function.
   *
   * @param {string} name Name of the bus
   * @param {InToOut<I, O>} decode decode function for this bus
   */
  static create<I, O>(
    name: string,
    decode: IODecode<I, O> | IOAsyncDecode<I, O>
  ): Bus<I, O> {
    return new Bus(name, async (input: I) => decode(input));
  }

  /**
   * Creates a new bus with the given name and bus, returning the
   * result of the first bus if it succeeds, otherwise the result of the second bus.
   *
   * On failure, returns the errors of both busses.
   *
   * @param {name} name The name of the new bus
   * @param {Bus<IB, OB>} other The other bus to combine with
   *
   * @returns {Bus<I | IB, O | OB>} A new bus "this || other"
   */
  public else<IB, OB>(
    other: Bus<IB, OB>,
    name: string = mergeNames(Tuple2.of(this.name, other.name), "||")
  ): Bus<I | IB, O | OB> {
    return new Bus<I | IB, O | OB>(name, async (input: I | IB) => {
      const us = await this.decode(input as I);

      if (us.isRight()) {
        return us;
      }

      const them = await other.decode(input as IB);

      if (them.isRight()) {
        return them;
      }

      return mergeLeft(us, them);
    });
  }

  /**
   *
   * @param {string} name The name of the new bus
   * @param {other} other The other bus to combine with
   *
   * @returns {Bus<I, O>} A new bus "this -> other"
   */
  public chain<OB>(
    other: Bus<O, OB>,
    name: string = mergeNames(Tuple2.of(this.name, other.name), "->")
  ): Bus<I, OB> {
    return new Bus(name, (input: I) =>
      this.decode(input).then((intermediate) =>
        intermediate.isLeft() ? intermediate : other.decode(intermediate.get())
      )
    );
  }
}
