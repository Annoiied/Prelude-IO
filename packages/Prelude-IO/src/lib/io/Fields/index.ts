export { default as number } from "./numberField";
