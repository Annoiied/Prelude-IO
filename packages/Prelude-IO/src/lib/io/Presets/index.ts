export { default as number } from "./number";
export { default as string } from "./string";
export { default as validNumber } from "./validNumber";
export { default as positiveNumber } from "./positiveNumber";
export { default as Optional } from "./Optional";
export { default as Vector } from "./Vector";
export { default as Complex } from "./Complex";
export { default as any } from "./any";
export { default as boolean } from "./boolean";
