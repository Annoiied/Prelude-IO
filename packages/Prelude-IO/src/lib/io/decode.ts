// import { Bus } from "./types";

// export default function decode<I, O, B = Bus<I, O>>(bus: B, input: I): O {
//   // TODO: Implement
// }
