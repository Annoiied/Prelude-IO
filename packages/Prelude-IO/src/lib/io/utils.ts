import { Either, Vector } from "prelude-ts";
import { IOError, IOErrors, IOLeft, IORight } from "./types";

const operators = ["AND", "OR", "->", "|"];

/**
 * Wraps a name in parentheses if it contains an operator.
 * Might need improvements in the future regarding presedence and pre-wrapped names.
 *
 * @param {string} name The name to check for operators
 * @returns {string} The name wrapped in parentheses if it contains an operator
 */
export const maybeWrapName = (name: string) =>
  operators.some((operator) => name.includes(operator)) &&
  (!name.startsWith("(") || !name.endsWith(")"))
    ? `(${name})`
    : name;

/**
 * Merges the names using -> or OR depending on the mode.
 *
 * Wraps the names in parentheses if they contain an operator.
 *
 * @example mergeNames(["a", "b"], "->") === "a -> b"
 * @example mergeNames(["a", "b -> c"], "OR") === "a OR (b -> c)"
 *
 * @param {[string, string]} names A tuple of names to merge
 * @param {string} operator The operator used for merging
 *
 * @returns {string} The merged names
 */
export const mergeNames = (names: [string, string], operator: string) =>
  names.map(maybeWrapName).join(` ${operator} `);

/**
 * Shorthand function for creating a right Either for IOResult.
 *
 * @param {O} output
 *
 * @returns {IORight<O>}
 */
export const IOAccept = <O>(output: O) => Either.right(output) as IORight<O>;

/**
 * Shorthand function for creating a left Either for IOResult.
 *
 *
 * @returns {IOLeft}
 */
export const IOReject = (error: IOErrors | IOError) =>
  Either.left(error instanceof Vector ? error : Vector.of(error)) as IOLeft;
