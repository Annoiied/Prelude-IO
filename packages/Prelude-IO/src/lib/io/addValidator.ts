import { Vector } from "prelude-ts";
import { VALIDATORS_KEY } from "../constants";
import IO from "./IO";
import { Validator } from "./Validators";

export default function addValidator(
  target: IO,
  field: string,
  validator: Validator
) {
  const validators =
    Reflect.getOwnMetadata(VALIDATORS_KEY, target, field) || Vector.empty();

  Reflect.defineMetadata(
    VALIDATORS_KEY,
    validators.append(validator),
    target,
    field
  );
}
