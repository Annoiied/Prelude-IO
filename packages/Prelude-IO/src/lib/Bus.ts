import { Option, Predicate } from "prelude-ts";
import { IOTransformer, IOAsyncTransformer } from "./types";
import { IOReject, mergeNames } from "./utils";

const chainTransformers =
  <T1, T2, T3>(
    name: string,
    a: IOAsyncTransformer<T1, T2>,
    b: IOAsyncTransformer<T2, T3>
  ): IOAsyncTransformer<T1, T3> =>
  (input: T1) =>
    a(input).then((intermediate) =>
      intermediate.isLeft()
        ? IOReject({
            condition: name,
            value: input,
            branches: intermediate.getLeft(),
          })
        : b(intermediate.get())
    );

const elseTransformers =
  <IA, OA, IB, OB>(
    name: string,
    a: IOAsyncTransformer<IA, OA>,
    b: IOAsyncTransformer<IB, OB>
  ) =>
  async (input: IA | IB) => {
    const us = await a(input as IA);

    if (us.isRight()) {
      return us;
    }

    const them = await b(input as IB);

    if (them.isRight()) {
      return them;
    }

    return IOReject({
      condition: name,
      value: input,
      branches: us.getLeft().appendAll(them.getLeft()),
    });
  };

/**
 * A bus is a wrapper around two transformers (serialize and deserialize), with some basic logic included such as chaining or unions.
 *
 * Busses are executed in parallel, so the order of the errors is not guaranteed (but should not matter).
 * When chaining busses, the deserialized value of the first bus is returned, the others are ignored. This might change in the future.
 *
 * Transformers do not need to be asyncronous on initiation for ease of use, but on Bus creation, they will
 * always be wrapped in a Promise.
 *
 * Transformers should ALWAYS resolve to a Result, never throw an error or reject. This is a philosophical choice based on
 * functional programming paradigms. Result objects allow functions to explicitly return errors, and the caller can decide
 * what to do with them. Throwing errors is a side effect, and should be avoided.
 *
 * [Predicates](http://emmanueltouzery.github.io/prelude.ts/latest/apidoc/files/predicate.html) can be attached to ensure
 * validity of the external type, that isn't relevant for the transforming process.
 *
 * @example
 * Ensure something is a number:
 * ```typescript
 * [[include:bus_creation_example.ts]]
 * ```
 *
 * @example
 * Add a predicate to a bus:
 * ```typescript
 * [[include:predicate_attachment_example.ts]]
 * ```
 *
 * @param <I> - The input type of the bus
 * @param <O> - The output type of the bus
 */
export default class Bus<I = unknown, O = unknown> {
  /**
   * Constructor cannot be called directly. Instead, {@link Bus.create} should be used.
   *
   * @internal
   */
  private constructor(
    /** The name used to identify this bus in errors */
    public readonly name: string,
    /** An optional [Predicate](http://emmanueltouzery.github.io/prelude.ts/latest/apidoc/files/predicate.html) exectuted on the output of the bus */
    public readonly predicate: Option<Predicate<O>>,
    /** A async deserialization function */
    public readonly deserialize: IOAsyncTransformer<I, O>,
    /** A async serialization function */
    public readonly serialize: IOAsyncTransformer<O, I>
  ) {
    Object.freeze(this);
  }

  /**
   * Creates a new bus with the given name and deserializer. Wraps the deserializer in an async function.
   *
   * @param <I> - The input type for the new bus
   * @param <O> - The output type for the new bus
   * @param name - Name of the bus
   * @param deserialize - deserialize function for this bus. Will be turned into a Promise.
   */
  static create<I, O>(
    name: string,
    deserialize: IOTransformer<I, O> | IOAsyncTransformer<I, O>,
    serialize: IOTransformer<O, I> | IOAsyncTransformer<O, I>
  ): Bus<I, O> {
    const fallbackRejection = (input: unknown, e: unknown) =>
      IOReject({
        condition: name,
        value: input,
        message: `Unexpected error: ${e}\n Busses should never throw errors, but instead return a Result!`,
      });

    return new Bus(
      name,
      Option.none(),
      async (input: I) => {
        try {
          return await deserialize(input);
        } catch (e) {
          return fallbackRejection(input, e);
        }
      },
      async (input: O) => {
        try {
          return await serialize(input);
        } catch (e) {
          return fallbackRejection(input, e);
        }
      }
    );
  }

  /**
   * Creates a new bus with the given name and bus, returning the
   * result of the first bus if it succeeds, otherwise the result of the second bus.
   *
   * On failure, returns the errors of both busses.
   *
   * @param <IB> - The other bus' input
   * @param <OB> - The other bus' output
   * @param name - The name of the new bus
   * @param other - The other bus to combine with
   *
   * @returns A new bus `this | other`
   */
  public else<IB, OB>(
    other: Bus<IB, OB>,
    name: string = mergeNames([this.name, other.name], "|")
  ): Bus<I | IB, O | OB> {
    return new Bus<I | IB, O | OB>(
      name,
      Option.none(),
      elseTransformers(name, this.deserialize, other.deserialize),
      elseTransformers(name, this.serialize, other.serialize)
    );
  }

  /**
   * Creates a new bus with the given name and bus, returning the
   * result of the first bus chained into the second bus.
   *
   * The predicates of the chained busses are preserved, but the newly created bus does not
   * have a predicate of its own.
   *
   * @param <OB> - The other bus' output
   * @param name - The name of the new bus
   * @param other - The other bus to combine with
   *
   * @returns A new bus "this -\> other"
   */
  public chain<OB>(
    other: Bus<O, OB>,
    name: string = mergeNames([this.name, other.name], "->")
  ): Bus<I, OB> {
    return new Bus(
      name,
      Option.none(),
      chainTransformers(name, this.deserialize, other.deserialize),
      chainTransformers(name, other.serialize, this.serialize)
    );
  }

  /**
   * Attaches a [Predicate](http://emmanueltouzery.github.io/prelude.ts/latest/apidoc/files/predicate.html) to a bus.
   * The predicate is applied on the deserialized value before returning.
   * @example
   * Add a predicate to a bus:
   * ```typescript
   * [[include:predicate_attachment_example.ts]]
   * ```
   *
   * Because this nests the original bus, you can call this function multiple times:
   * ```typescript
   * [[include:multiple_predicates_example.ts]]
   * ```
   *
   * @param name - The name of the predicate. Formatted to be "predicateName(busName)"
   * @param predicate - The predicate to attach to this bus
   *
   * @returns A new bus with the given predicate attached to it.
   */
  public if(name: string, predicate: Predicate<O>): Bus<I, O> {
    const newBusName = `${name}(${this.name})`;

    const newserialize = async (input: O) => {
      if (!predicate(input)) {
        return IOReject({
          condition: newBusName,
          value: input,
          branches: IOReject({
            condition: name,
            value: input,
          }).getLeft(),
        });
      }

      return this.serialize(input);
    };

    const newDeserialize = async (input: I) => {
      const deserialized = await this.deserialize(input);

      if (deserialized.isLeft()) {
        return deserialized;
      }

      if (predicate(deserialized.get())) {
        return deserialized;
      }

      return IOReject({
        condition: newBusName,
        value: input,
        branches: IOReject({
          condition: name,
          value: deserialized.get(),
        }).getLeft(),
      });
    };

    return new Bus(
      newBusName,
      Option.of(predicate),
      newDeserialize,
      newserialize
    );
  }
}
