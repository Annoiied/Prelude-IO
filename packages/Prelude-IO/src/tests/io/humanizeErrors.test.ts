import * as io from "../../lib/io";

describe("io.humanizeErrors", () => {
  const Cat = io.Complex("Cat", {
    name: io.string,
    age: io.positiveNumber,
    isAdopted: io.boolean,
  });

  const Litter = io.Complex("Litter", {
    birthplace: io.string,
    mother: Cat,
    // Questions may be directed to the author's therapist.
    father: io.Optional(Cat),
    kittens: io.Vector(Cat),
  });

  it("returns a string of more readable errors", async () => {
    const input = {
      birthplace: false, // They're hiding that they were born on the Moon
      mother: {
        name: "Mittens",
        age: 5,
        isAdopted: true,
      },
      father: null,
      kittens: [
        {
          name: "Suku",
          age: 1,
          isAdopted: false,
        },
        {
          name: "Skadi",
          age: 1,
          isAdopted: false,
        },
        {
          name: "Sif",
          // This kitten is a work in progress
          age: -1,
          isAdopted: "bit too soon",
        },
      ],
    };

    const fullErrors = (await Litter.deserialize(input)).getLeftOrThrow();

    expect(io.humanizeErrors(fullErrors)).toMatchSnapshot();
  });
});
