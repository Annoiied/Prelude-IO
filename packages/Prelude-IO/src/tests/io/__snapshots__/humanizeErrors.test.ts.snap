// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`io.humanizeErrors returns a string of more readable errors 1`] = `
"Litter
  ├─ birthplace
  │   └─ isString(any)
  │       └─ isString rejected the value \`false\`
  └─ kittens
      └─ Vector(Cat)
          └─ [2]
              └─ Cat
                  ├─ isAdopted
                  │   └─ isBoolean(any)
                  │       └─ isBoolean rejected the value \`bit too soon\`
                  └─ age
                      └─ isPositive(isValidNumber(isNumber(any)))
                          └─ isPositive rejected the value \`-1\`"
`;
